### PR TITLE
Fix #389 Adding unit test for install-cli via proxy

### LINKED ISSUE #389: Support the use of install-cli in HTTP proxy environment
**OS details:**
Windows 7

**Provider:**
VirtualBox

**Output of `vagrant plugin list`:**

```
vagrant-registration (1.3.0)
vagrant-service-manager (1.3.3)
```

**Steps to reproduce the issue:**
1.  A Windows proxy environment. Only web browser has access to internet with proxy.
2. vagrant service-manager install-cli openshift

`vagrant service-manager install-cli openshift` should download and install oc binary. At this point the workstation's CMD or Cygwin does not have access to internet through proxy.  So VSM should setup the proxy for the Vagrant environment and then use it to download stuff for `install-cli` feature.

### PR DESCRIPTION
**Pull request by hferentschik**
Fix #389

### CODE REVIEW

**Comment by hferentschik** [comment]
@budhrg here we go. Please review.

File: test/vagrant-service-manager/proxy_test.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+        it 'client binary download request should pass through proxy' do
+          @http_proxy.expect_http_connect('github.com:443')
+          dirname = File.dirname(__FILE__)
+          filename = File.join(dirname, '..', 'test_data', URI(@handler.url).path.split('/').last.to_s)
+          @https_server.expect_and_respond(@handler.url, filename)
+
+          @handler.download_archive
+
+          @http_proxy.assert_connect_requests.must_equal true
```
Comment (brgnepal) [inline test/vagrant-service-manager/proxy_test.rb:51]: :+1:

File: test/vagrant-service-manager/proxy_test.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        before do
+          @http_proxy = FakeHTTPProxyServer.start ServiceManager.temp_dir
+          @https_server = FakeHTTPSServer.start
+
+          @machine.config.servicemanager.proxy = "http://localhost:#{FakeHTTPProxyServer::DEFAULT_PORT}"
+        end
+
+        it 'client binary download request should pass through proxy' do
+          @http_proxy.expect_http_connect('github.com:443')
+          dirname = File.dirname(__FILE__)
```
Comment (brgnepal) [inline test/vagrant-service-manager/proxy_test.rb]: @hferentschik I think we can take `dirname` at top level `before` block. It seems to be common to both cases.

File: test/vagrant-service-manager/proxy_test.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+          filename = File.join(dirname, '..', 'test_data', URI(@handler.url).path.split('/').last.to_s)
+          @https_server.expect_and_respond(@handler.url, filename)
```
Comment (brgnepal) [inline test/vagrant-service-manager/proxy_test.rb]: So, we are kind of making the stage here by filling `@url_request_received` by calling this line, right?

File: test/vagrant-service-manager/proxy_test.rb (hunk 2 above)
Comment (hferentschik) [reply test/vagrant-service-manager/proxy_test.rb]: done

File: test/vagrant-service-manager/proxy_test.rb (hunk 3 above)
Comment (hferentschik) [reply test/vagrant-service-manager/proxy_test.rb]: we are telling the server which urls to expect and how to respond to it. You can call this setting the stage, yes.

File: test/vagrant-service-manager/proxy_test.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      describe 'Unauthenticated proxy environment' do
+        before do
+          @http_proxy = FakeHTTPProxyServer.start ServiceManager.temp_dir
+          @https_server = FakeHTTPSServer.start
+
+          @machine.config.servicemanager.proxy = "http://localhost:#{FakeHTTPProxyServer::DEFAULT_PORT}"
+        end
+
+        it 'client binary download request should pass through proxy' do
+          @http_proxy.expect_http_connect('github.com:443')
```
Comment (brgnepal) [inline test/vagrant-service-manager/proxy_test.rb:46]: @hferentschik No idea what this line is doing.

Comment (hferentschik) [reply test/vagrant-service-manager/proxy_test.rb:46]: testing the proxy. The proxy does not see a url, only the HTTPS server. We want to make sure the connection is established via the proxy (that was our primary goal, right?). We then also confirm that the HTTPS receives the actual download URL and sends a faked response.

**Comment by brgnepal** [comment]
LGTM :+1: 

Tests are passing too.

File: test/vagrant-service-manager/proxy_test.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+          @https_server.expect_and_respond(@handler.url, @filename)
+
+          @handler.download_archive
+
+          @http_proxy.assert_connect_requests.must_equal true
+          @https_server.assert_requests.must_equal true
+        end
+      end
+
+      describe 'Authenticated proxy environmentt' do
```
Comment (brgnepal) [inline test/vagrant-service-manager/proxy_test.rb]: Typo error here in environment.

File: test/fake_https_server.rb [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    @https_server
+  end
+
+  def self.stop
+    @https_server.shutdown unless @https_server.nil?
+  end
+
+  def initialize(port)
+    # we need to temporarily capture stderr, since WEBrick::Utils::create_self_signed_cert prints to stderr
+    capture_stderr do
```
Comment (brgnepal) [inline test/fake_https_server.rb:22]: :+1: Nice one

**Comment by brgnepal** [comment]
Merged and thanks for contribution.

**Pull request closed without merging**